platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+	}
+
+	if (data->num_input_seg > mdl->info.num_inputs &&
+	    (_odp_ml_glb->capa.max_segs_per_input == 1)) {
+		_ODP_ERR("Segmented input data is not supported\n");
+		return -1;
+	}
+
+	if (data->num_output_seg > mdl->info.num_outputs &&
+	    (_odp_ml_glb->capa.max_segs_per_output == 1)) {
+		_ODP_ERR("Segmented output data is not supported");
+		return -1;
+	}
+
+	uint32_t size = 0;
+	uint32_t input_index = 0;
+	uint32_t seg_size_sum = 0;
+	odp_bool_t index_new = true;
+	uint32_t segs_per_input = 1;
+
+	for (uint32_t i = 0; i < data->num_input_seg; i++) {
+		if (data->input_seg[i].addr == NULL) {
+			_ODP_ERR("data->input_seg[%u].addr must not NULL\n", i);
+			return -1;
+		};
+
+		if (index_new) {
+			if (input_index > mdl->info.num_inputs - 1) {
+				_ODP_ERR("Too much number of input segments given\n");
+				return -1;
+			}
+
+			/* Input with dynamic batch size */
+			if (mdl->input_info[input_index].shape.type == ODP_ML_SHAPE_BATCH)
+				size = dyn_io_size(&mdl->input_info[input_index].shape,
+						   mdl->input_info[input_index].data_type_size,
+						   param);
+			else
+				size = mdl->input_sizes[input_index];
+
+			if (!size) {
+				_ODP_ERR("Size for %uth input is 0\n", input_index);
+				return -1;
+			}
+		}
+
+		seg_size_sum += data->input_seg[i].size;
+
+		if (seg_size_sum > size) {
+			_ODP_ERR("Sum of segment sizes %u exceeds %uth input data size %u\n",
+				 seg_size_sum, input_index, size);
+			return -1;
+		}
+
+		if (seg_size_sum == size) {
+			if (segs_per_input > _odp_ml_glb->capa.max_segs_per_input) {
+				_ODP_ERR("Number of segments %u for input[%u] exceeds maximum"
+					 " number of data segments per model input %u\n",
+					 segs_per_input, input_index,
+					 _odp_ml_glb->capa.max_segs_per_input);
+				return -1;
+			}
+			input_index++;
+			index_new = true;
+			seg_size_sum = 0;
+			segs_per_input = 1;
+		} else {
+			segs_per_input++;
+			index_new = false;
+		}
+	}
+
+	if (input_index != mdl->info.num_inputs) {
+		_ODP_ERR("Data is not provided for all model inputs\n");
+		return -1;
+	}
+
+	seg_size_sum = 0;
+	index_new = true;
+	uint32_t output_index = 0;
+	uint32_t segs_per_output = 1;
+
+	for (uint32_t i = 0; i < data->num_output_seg; i++) {
+		if (data->output_seg[i].addr == NULL) {
+			_ODP_ERR("data->output_seg[%u].addr must not NULL\n", i);
+			return -1;
+		}
+
+		if (index_new) {
+			if (output_index > mdl->info.num_outputs - 1) {
+				_ODP_ERR("Too much number of output segments given\n");
+				return -1;
+			}
+
+			/* Output with dynamic batch size */
+			if (mdl->output_info[output_index].shape.type == ODP_ML_SHAPE_BATCH)
+				size = dyn_io_size(&mdl->output_info[output_index].shape,
+						   mdl->output_info[output_index].data_type_size,
+						   param);
+			else
+				size = mdl->output_sizes[output_index];
+
+			if (!size) {
+				_ODP_ERR("Size for %uth output is 0\n", output_index);
+				return -1;
+			}
+		}
+
+		seg_size_sum += data->output_seg[i].size;
+
+		if (seg_size_sum >= size) {
```

Comment:
Isn't (seg_size_sum > sum) an error also on output side (it's reported on input side)